src/schedule_synthesis.jl

```diff
@@ -0,0 +1,273 @@
+    # scheduler automaton
+    current_states = Dict{Int64, LinkedList{Int64}}(AS.l_0 => list(AS.l_0))
+
+    # Traverse the tree until 
+    #   (1) the required number of steps is reached,
+    #   (2) a cycle is found, or 
+    #   (3) no more states are valid for exploration
+    for step in 1:H
+        next_states = Dict{Int64, LinkedList{Int64}}()
+        for (l, path) in pairs(current_states), σ in AS.Σ
+            l_new = AS.T(l, σ)
+            if !AS.Q_f(l_new) || haskey(next_states, l_new)
+                continue
+            elseif l_new in path
+                # Found cycle -> Case (2)
+                return _path_to_schedule(reverse(cons(l_new, path)), AS)
+            end
+            next_states[l_new] = cons(l_new, path)
+        end
+
+        if isempty(next_states)
+            # No more valid states -> Case (3)
+            return Vector{Int64}()
+        end
+
+        current_states = next_states
+    end
+
+    for (l, path) in pairs(current_states)
+        if AS.Q_f(l)
+            # If the outer loop ends and accepting state is found -> Case (1)
+            return _path_to_schedule(reverse(path), AS)
+        end
+    end
+
+    # If the outer loop ends and accepting state is not found -> Case (1)
+    return Vector{Int64}()
+end
+
+"""
+    function synthesize_constraints(sysd, K, z_0, d_max, maxwindow, n, t)
+
+Find all `MeetAny` weakly hard constraints with window size at most `maxwindow` that guarantees the deviation upper bound is at most `d_max`. The system is specified by [`Automaton`](@ref) `a` and initial state is `z_0`. `n` and `t` are as in [`bounded_runs_iter`](@ref).
+"""
+function synthesize_constraints(sysd::AbstractStateSpace{<:Discrete},
+    K::AbstractMatrix{Float64}, z_0::AbstractVecOrMat, d_max::Float64,
+    maxwindow::Int64, n::Int64, t::Int64)::Vector{WeaklyHardConstraint}
+
+    # TODO: 
+    #   1) Test to make sure it works the same
+    #   2) Discuss: should additional safe constraints be included in result? 
+    #      Or only the ones that are actually checked & useful for scheduling
+    safe_constraints = MeetAny[]
+
+    # Do not need to go through all O(maxwindow^2) constraints,
+    # see paper for optimization argument
+    meet = 1
+    for window in 2:maxwindow
+        while meet < window
+            constraint = MeetAny(meet, window)
+            a = hold_kill(sysd, K, constraint)
+            # Check if the deviation bound is within the safety margin
+            reachable = bounded_runs_iter(a, z_0, n, t)
+            m = maximum(deviation(a, z_0, reachable))
+            if m <= d_max
+                # All constraints with (m, window) where m >= meet are valid
+                push!(safe_constraints, constraint)
+                break
+            end
+            meet += 1
+        end
+    end
+
+    safe_constraints
+end
+
+# Helper functions
+
+"""
+    _undigit(d[, base=2])
+
+Convert a list of digits to a number. Default base=2.
+```julia
```

Comment:
Might as well make this block a [`jldoctest`](https://documenter.juliadocs.org/stable/man/doctests/) block so it can be automatically run as a test.